net/interfaces/interfaces.go

```diff
@@ -83,6 +83,14 @@ func isProblematicInterface(nif *net.Interface) bool {
 // LocalAddresses returns the machine's IP addresses, separated by
 // whether they're loopback addresses.
 func LocalAddresses() (regular, loopback []netaddr.IP, err error) {
+	regular, loopback, err = localAddressesMaybeLinkLocal(false)
+	if len(regular) > 0 {
+		return
+	}
+	return localAddressesMaybeLinkLocal(true)
```

Comment:
this means that on Lamba/etc, we end up calling net.Interfaces() etc twice, which is a little bit wasteful.

But mostly I'm not a huge fan of boolean parameters. Reading code later, a callsite's random foo(false) or foo(true) is always mysterious.

Also: this function already returns `(regular, loopback)` for magicsock to only use `regular` if it's non-empty, else it falls back to using loopback. But now there's yet another mechanism for "use class X of addresses unless it's empty, else use something different". Ideally both would use the same mechanism. Really the logic is: return our addresses, preferring first real IPs, else link local, else loopbacks.

But then magicsock also wants the type, IIRC. So maybe the signature of `func LocalAddresses` should be `func LocalAddresses() (addrs []netaddr.IP, addrType T, err error)` and its body can only call `net.Interfaces` once and do the selection of best addresses and remove that from magicsock?

Reply:
wgengine/magicsock/magicsock.go will only use loopback if there are no public endpoint addresses:

```
                 if len(ips) == 0 && len(eps) == 0 {
                         ips = loopback
```

I don't think we can move that part of the functionality here, net/interfaces doesn't know about external addresses.

This routine could return three groups of interfaces (prime, subprime, junk) and let MagicSock choose.

Maybe `(regular, linklocal, loopback []netaddr.IP, err error)` instead of making a bond market joke.

Reply:
Ended up making interfaces.go decide whether to use regular vs linklocal, but letting magicsock decide whether to use localhost so it can consider whether external addresses exist.